dashbuilder-client/dashbuilder-widgets/src/main/java/org/dashbuilder/client/widgets/dataset/editor/DataSetDefPreviewTable.java

```diff
@@ -83,31 +87,71 @@ public void show(final DataSetDef dataSetDef, final Collection<DataColumnDef> co
         if (dataSetDef != null) {
             
             // Build the table displayer settings.
-            final TableDisplayerSettingsBuilder<TableDisplayerSettingsBuilderImpl> tableDisplayerSettingsBuilder = DisplayerSettingsFactory.newTableSettings()
+            final TableDisplayerSettingsBuilder settingsBuilder = DisplayerSettingsFactory.newTableSettings()
                     .dataset(dataSetDef.getUUID())
                     .renderer(DefaultRenderer.UUID)
                     .titleVisible(false)
                     .tablePageSize(6)
                     .tableOrderEnabled(true)
                     .filterOn(true, false, false);
 
+            // Make only the specified columns visible (if null then show all)
             if (columns != null && !columns.isEmpty()) {
                 for (final DataColumnDef column : columns) {
-                    tableDisplayerSettingsBuilder.column(column.getId());
+                    settingsBuilder.column(column.getId());
                 }
             }
 
-            final DisplayerSettings settings = tableDisplayerSettingsBuilder.buildSettings();
             // Disable backend cache for preview.
             DataSetDef editCloneWithoutCacheSettings = dataSetDef.clone();
             editCloneWithoutCacheSettings.setCacheEnabled(false);
 
+            // Close the current displayer (if any)
+            if (tableDisplayer != null) {
+                tableDisplayer.close();
+            }
+
             // Configure the table displayer and the data set handler for edition.
+            DisplayerSettings settings = settingsBuilder.buildSettings();
             tableDisplayer = displayerLocator.lookupDisplayer(settings);
             tableDisplayer.setDataSetHandler(new DataSetEditHandler(clientServices, settings.getDataSetLookup(), editCloneWithoutCacheSettings));
+
+            // For CSV datasets, make sure dates & numbers are displayed as defined in the CSV configuration
+            if (dataSetDef instanceof CSVDataSetDef) {
+                CSVDataSetDef csvDataSetDef = (CSVDataSetDef) dataSetDef;
+                tableDisplayer.addListener(new AbstractDisplayerListener() {
+
+                    // The column settings are set right after the data set load, once the dataset columns are known
+                    @Override
+                    public void onDataLoaded(Displayer displayer) {
+                        configureColumnSettings(displayer, csvDataSetDef);
+                    }
+                });
+            }
+
+            // Draw the table
             draw(displayerListener);
         }
+    }
+
+    protected void configureColumnSettings(Displayer displayer, CSVDataSetDef csvDataSetDef) {
 
+        DataSet dataSet = displayer.getDataSetHandler().getLastDataSet();
+        for (DataColumn column : dataSet.getColumns()) {
```

Comment:
@dgutierr - same here, try to avoid external iterations as soon as possible and do an incremental replacement by internal iterators  / use of streams...